### PR TITLE
BUG: extra leading space in to_string when index=False

### LINKED ISSUE #24980: Bug on .to_string(index=False)
#### Code Sample, a copy-pastable example if possible
Related issue (very old though) #11833

```python
import pandas

print("pandas.__version__ == {:s}".format(pandas.__version__))

columns = ['name', 'age']
values = [
    ['name_one', '31'],
    ['name_two', '32']]
data_frame = pandas.DataFrame(values, columns=columns)

# filter

filtered = data_frame[(data_frame['name'] == 'name_one') & (data_frame['age'] == '31')]
# single value comes with a blank at the beginning when running 0.24.0

print("==={:s}===".format(filtered.name.to_string(index=False)))

```
The output of this running on pandas 0.24.0 is:
(I put this in a file called ``pandasbug.py``)

```shell 
(goodman_pipeline) [simon@ctioy9 sandbox]$ python3.7 pandasbug.py 
pandas.__version__ == 0.24.0
=== name_one===
```

#### Problem description

Exactly when ``0.24.0`` was released my automatic builds started to fail. The problem, a file not found.
This was erroneously reported by me on astropy/ccdproc#658 
I'm using a pandas data frame to filter a set of reference files and when extracting the file name and join it to a full path it turns out a non-existing path such as:
``/full/path/to/[unwanted-blank-here]file_name.fits`` (``/full/path/to/ file_name.fits``)

[This travis build](https://travis-ci.org/soar-telescope/goodman_pipeline/jobs/485402650) will show the real-world error caused, if you can access it.

#### Expected Output
In my system I have pandas 0.20.3 working with python 3.6 and the latests working build ran pandas ``0.23.4``

```shell
(goodman_pipeline) [simon@ctioy9 sandbox]$ python3.6 pandasbug.py 
pandas.__version__ == 0.20.3
===name_one===
```

#### Output of ``pd.show_versions()``

<details>
>>> import pandas 
>>> pandas.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-1.el7.elrepo.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.0
pytest: 4.1.1
pip: 18.1
setuptools: 40.6.3
Cython: 0.29.2
numpy: 1.15.2
scipy: 1.2.0
pyarrow: None
xarray: None
IPython: None
sphinx: 1.8.3
patsy: None
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [x] closes #24980 , xref #28538
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I still see people posting issue for the change of behaviour in `to_string(index=False)` on pandas such as #28538 , so I think it might be worth it to reopen my previous stalled PR #25000 and see if could get it solved this time.

All the changes have been made based on the reviews in PR #25000 and for detailed summary and explanation, please check https://github.com/pandas-dev/pandas/pull/25000#issuecomment-458714951
Feel free to take a look and any review and comments are very welcomed!

### CODE REVIEW

File: pandas/io/formats/format.py [hunk 1]
```diff
@@ -1095,7 +1106,7 @@ def format_array(
     space: Optional[Union[str, int]] = None,
     justify: str = "right",
     decimal: str = ".",
-    leading_space: Optional[bool] = None,
+    leading_space: bool = "compat",
```
Comment (alimcmaster1) [inline pandas/io/formats/format.py]: this isn't a boolean? Can you run mypy on your changes

File: pandas/tests/io/formats/test_format.py [hunk 2]
```diff
@@ -3272,3 +3272,43 @@ def test_filepath_or_buffer_bad_arg_raises(float_frame, method):
     msg = "buf is not a file name and it has no write method"
     with pytest.raises(TypeError, match=msg):
         getattr(float_frame, method)(buf=object())
+
+
+@pytest.mark.parametrize(
+    "inputs, expected",
+    [([" a", " b"], " a\n b"), ([".1", "1"], ".1\n 1"), (["10", "-10"], " 10\n-10")],
+)
+def test_to_string_index_false_corner_case(inputs, expected):
```
Comment (alimcmaster1) [inline pandas/tests/io/formats/test_format.py]: I think these cases should just be added as param's to your test method below `test_format_remove_leading_space_series`

File: pandas/io/formats/format.py [hunk 3]
```diff
@@ -937,13 +943,18 @@ def to_latex(
     def _format_col(self, i: int) -> List[str]:
         frame = self.tr_frame
         formatter = self._get_formatter(i)
+        if self.index:
+            leading_space = "compat"
```
Comment (alimcmaster1) [inline pandas/io/formats/format.py]: Seems like there was already some discussion around this here https://github.com/pandas-dev/pandas/pull/25000/files#r252237505 -

File: pandas/io/formats/format.py [hunk 4]
```diff
@@ -1110,7 +1121,7 @@ def format_array(
     space
     justify
     decimal
-    leading_space : bool, optional
+    leading_space : bool, default is 'compat'
         Whether the array should be formatted with a leading space.
```
Comment (alimcmaster1) [inline pandas/io/formats/format.py]: same comment as previous here what is the intended type of this?

**Review by alimcmaster1 — CHANGES_REQUESTED** [review]

**Comment by alimcmaster1** [comment]
thanks for the pr @charlesdong1991 - some comments above. We can re-open the previous PR if helpful?  #25000

File: pandas/io/formats/format.py (hunk 1 above)
Comment (charlesdong1991) [reply pandas/io/formats/format.py]: oh, yeah, didn't see your comment when I committed my latest change. I changed already

File: pandas/tests/io/formats/test_format.py (hunk 2 above)
Comment (charlesdong1991) [reply pandas/tests/io/formats/test_format.py]: sure, this was just some corner cases that was found during the previous PR. will move.

File: pandas/io/formats/format.py (hunk 4 above)
Comment (charlesdong1991) [reply pandas/io/formats/format.py]: yeah, there was no type hinting back then, and I didn't notice this so just copy paste the change in my previous PR. my bad, changed.

File: pandas/io/formats/format.py (hunk 3 above)
Comment (charlesdong1991) [reply pandas/io/formats/format.py]: emm, i think i addressed it already in this PR?

**Comment by charlesdong1991** [comment]
thanks for your comment, @alimcmaster1 
I already have this PR open and there has been dramatic changes since back then, so I think it is fine just to leave the old PR as is. But if you think it's helpful, feel free to re-open the stalled PR, and I don't think I am able to do it.

**Comment by charlesdong1991** [comment]
Finally the annotation error is gone, it's green now 😅 

pls take a look and see if this PR is useful.

**Comment by charlesdong1991** [comment]
Hi, @jreback 
I would appreciate a lot if you would also like to take a look at this PR which was closed long ago, but since other people posted the issue again, I reopened it with this PR. 

All comments in the last closed PR were addressed.

**Comment by simontorres** [comment]
So, this PR seems ok. Anyone with the right permissions to fix conflict an merge? I've been forced to keep using version `0.23.0`

**Comment by jreback** [comment]
@simontorres there are quite a number of comments on this
it is not so simple

i find it hard to believe this blocking an update of pandas

**Comment by simontorres** [comment]
I was talking out ignorance. I'm sorry

**Comment by charlesdong1991** [comment]
Hi, @simontorres i resolved the conflict, and as @jreback said, there are some comments in the PR, and pls feel free to take over if you could find a better solution for this, and then i could close this PR to make way for you!

**Comment by simontorres** [comment]
Thanks @charlesdong1991 and @jreback . I don't think I can find a better solution but I will see if I can help.

**Comment by charlesdong1991** [comment]
I am sure you could help fix the issue and pandas 👍  @simontorres 

also @jreback feel free to close this PR if you find it hard to approach!

**Comment by alimcmaster1** [comment]
> Thanks @charlesdong1991 and @jreback . I don't think I can find a better solution but I will see if I can help.

@simontorres 
Could you provide a repo of the issue you faced? That way we can ensure we are covering your use case with a test case in pandas.

**Comment by simontorres** [comment]
# Introduction/Application
One example is this Imaging Spectrograph called [GoodmanHTS](http://www.ctio.noao.edu/soar/content/goodman-high-throughput-spectrograph), there are two elements that can move a camera and a dispersor, their angle are recorded in the image's header. There are certain _named modes_ that I'm interested in that can be used define file names, for instance. So this is a shorter version of a longer table.

```python
import pandas

columns = ['grating_freq', 'wavmode', 'camtarg', 'grttarg', 'ob_filter']
spec_mode = [['400', 'm1', '11.6', '5.8', 'None'],
             ['400', 'm2', '16.1', '7.5', 'GG455'],
             ['600', 'UV', '15.25', '7.0', 'None'],
             ['600', 'Blue', '17.0', '7.0', 'None'],
             ['600', 'Mid', '20.0', '10.0', 'GG385'],
             ['600', 'Red', '27.0', '12.0', 'GG495'],
             ['930', 'm1', '20.6', '10.3', 'None'],
             ['930', 'm2', '25.2', '12.6', 'None'],
             ['930', 'm3', '29.9', '15.0', 'GG385'],
             ['930', 'm4', '34.6', '18.3', 'GG495'],
             ['930', 'm5', '39.4', '19.7', 'GG495']]

modes_data_frame = pandas.DataFrame(spec_mode, columns=columns)
```

# The problem

The information I get is the Grating Frequency (for instance 930), the Camera Target Angle (camtarg, 20.6), the  Grating Target Angle (grttarg, 10.3) and the Order Blocking Filter (. When I use pandas to find the match in the following way.

```python
mode = modes_data_frame[
    ((modes_data_frame['grating_freq'] == 930) &
     (modes_data_frame['camtarg'] == 20.6) &
     (modes_data_frame['grttarg'] == 10.3) &
     (modes_data_frame['ob_filter'] == None))]

if mode.empty:
    central_wavelength = get_central_wavelength(*some_args)
    return 'Custom_{:d}nm'.format(central_wavelength)
else:
    return mode['wavmode'].to_string(index=False)
```

it should return 'm1' instead it returns ' m1'
This worked until version 0.23.0 and I reported it on #24980

**Comment by WillAyd** [comment]
This hasn't gotten traction in a while so closing as stale, but @charlesdong1991 or @simontorres ping if you plan to pick up and can continue from there
